adaptdl/adaptdl/torch/data.py

```diff
@@ -269,41 +269,38 @@ def autoscale_batch_size(self, max_batch_size, local_bsz_bounds=None,
+        if self.max_batch_size is None or goodput_fn is None:
             # No autoscale batch size, just divide batch size evenly.
             self._current_local_bsz = math.ceil(self.batch_size /
                                                 adaptdl.env.num_replicas())
             self._accumulation_steps = 0
-        else:
-            # Autoscale batch size, compute on rank 0 and broadcast.
-            speedup_fn = get_speedup_fn()
-
+        elif self._current_local_bsz is None:
             # if init, use the batch size suggested
-            if self._current_local_bsz is None:
-                _, (local_bsz, accumulation_steps) = speedup_fn(
-                    adaptdl.env.num_nodes(),
-                    adaptdl.env.num_replicas(),
-                    return_config=True)
-                self._current_local_bsz = local_bsz
-                self._accumulation_steps = accumulation_steps
-
+            _, atomic_bsz, accum_steps = goodput_fn.optimize(
+                adaptdl.env.num_nodes(), adaptdl.env.num_replicas(),
+                max_batch_size=self._max_batch_size,
+                atomic_bsz_range=self._local_bsz_bounds,
+                accumulation=self._gradient_accumulation)
+            self._current_local_bsz = atomic_bsz
+            self._accumulation_steps = accum_steps
+        else:
             # if not first time, we check against the relative speedup
-            else:
-                suggest_speedup, (local_bsz, accumulation_steps) = speedup_fn(
-                                   adaptdl.env.num_nodes(),
-                                   adaptdl.env.num_replicas(),
-                                   return_config=True)
-                # get current speedup
-                current_speedup = speedup_fn(adaptdl.env.num_nodes(),
-                                             adaptdl.env.num_replicas(),
-                                             local_bsz=self.current_local_bsz)
-                # use only if speedup is significant
-                speedup_to_cur = suggest_speedup / current_speedup
-                if speedup_to_cur > self.speedup_threshold:
-                    self._current_local_bsz = local_bsz
-                    self._accumulation_steps = accumulation_steps
-
-        (self._current_local_bsz, self._accumulation_steps) = \
+            suggest_goodput, atomic_bsz, accum_steps = goodput_fn.optimize(
+                adaptdl.env.num_nodes(), adaptdl.env.num_replicas(),
+                max_batch_size=self._max_batch_size,
+                atomic_bsz_range=self._local_bsz_bounds,
+                accumulation=self._gradient_accumulation)
+            # get current goodput
+            current_goodput = goodput_fn(
+                adaptdl.env.num_nodes(), adaptdl.env.num_replicas(),
+                self.current_local_bsz, self.accumulation_steps)
+            # use only if speedup is significant
+            if suggest_goodput / current_goodput > self.speedup_threshold:
```

Comment:
need to catch div by zero from `goodput_fn.optimize` not producing any valid batchsizes.